### DISCUSSION TITLE
Fixing ekg-core #41

### DISCUSSION THREAD

**Discussion by TravisWhitaker** [haskell-github-trust/.github#13] [General]
Greetings,

I am interested in finally resolving https://github.com/haskell-github-trust/ekg-core/issues/41 . As it stands, ekg is essentially unusable on aarch64 platforms because of an incorrect C implementation of atomic operations that happens to work on certain architectures (aarch64 is not one of them). When I opened this issue four years ago, my colleagues and I were probably among the only people who cared about Haskell on aarch64. Hopefully with the proliferation of Apple Silicon there are now more people in the wider community who are interested in things working well on aarch64.

Based on my measurements, the best (or at lest the best performance) way forward is to use GHC's primops to replace the C implementations:  https://github.com/haskell-github-trust/ekg-core/pull/42 . However, this PR didn't seem to gain any traction, so I'm curious to get others' opinions on the right way forward. I think what we should _not_ do is continue to leave the broken implementations in place, as the library will appear to work on aarch64 but will give incorrect results.

**Comment by jamesdbrock** [comment]
Hi @TravisWhitaker I sent you an ownership invitation for this org, so now you can merge PRs and publish that package.

@L0neGamer has been working on the **ekg** family of packages recently so maybe he has an opinion on that issue and would discuss it with you.

**Reply by L0neGamer** [reply]
I'm happy to review that pr upon rerun of benchmarks if that's feasible. Else I'll review when I have a moment.

**Reply by TravisWhitaker** [reply]
@jamesdbrock thank you for adding me, I'd be pleased to help.

@L0neGamer I have some free time next week (US holiday). I'll bring that PR forward to current master and clean up my benchmarks so they can be included.

**Reply by L0neGamer** [reply]
Hi @TravisWhitaker, have you made the changes you wanted to?

**Reply by TravisWhitaker** [reply]
@L0neGamer I have indeed (admittedly later than I planned). This should be ready to merge if you'd like.

Note that this does weaken the counter interface from `Int64` to `Int` (reason being the primops used can't be lowered on the supported 32-bit machines), so a major version bump may be necessary.
